### PR TITLE
Sanity checks & Prevent unecessary model resizing

### LINKED ISSUE #137: CUDA Error
### Code sample

```python
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, roc_auc_score

from datasets import Dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    TrainingArguments,
    Trainer
)
import torch


# =========================
# 1. Préparation des données
# =========================

X = df["text"].astype(str)
y = df["label"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

train_df = pd.DataFrame({"text": X_train, "label": y_train})
test_df  = pd.DataFrame({"text": X_test,  "label": y_test})

train_dataset = Dataset.from_pandas(train_df.reset_index(drop=True))
test_dataset  = Dataset.from_pandas(test_df.reset_index(drop=True))


# =========================
# 2. Tokenizer
# =========================

model_name = "camembert-base"  # adapté au français
tokenizer = AutoTokenizer.from_pretrained(model_name)

def tokenize(example):
    return tokenizer(
        example["text"],
        truncation=True,
        padding="max_length",
        max_length=256
    )

train_dataset = train_dataset.map(tokenize, batched=True)
test_dataset  = test_dataset.map(tokenize, batched=True)

train_dataset.set_format(type="torch", columns=["input_ids", "attention_mask", "label"])
test_dataset.set_format(type="torch", columns=["input_ids", "attention_mask", "label"])


# =========================
# 3. Modèle
# =========================

model = AutoModelForSequenceClassification.from_pretrained(
    model_name,
    num_labels=2
)


# =========================
# 4. Métriques
# =========================

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    probs = torch.softmax(torch.tensor(logits), dim=1)[:, 1].numpy()
    preds = np.argmax(logits, axis=1)

    return {
        "roc_auc": roc_auc_score(labels, probs),
        "f1": (preds == labels).mean()
    }


# =========================
# 5. Entraînement
# =========================

training_args = TrainingArguments(
    output_dir="./results",
    learning_rate=2e-5,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=3,
    eval_strategy="epoch",
    save_strategy="epoch",
    logging_dir="./logs",
    load_best_model_at_end=True
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    #tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

trainer.train()


# =========================
# 6. Évaluation finale
# =========================

predictions = trainer.predict(test_dataset)
logits = predictions.predictions
labels = predictions.label_ids

probs = torch.softmax(torch.tensor(logits), dim=1)[:, 1].numpy()
preds = np.argmax(logits, axis=1)

print(classification_report(labels, preds))
print("ROC AUC:", roc_auc_score(labels, probs))
```

Image

```python
import torch
from transformers import AutoModelForCausalLM, AutoModelForSequenceClassification, AutoTokenizer

from interpreto import (
    Granularity,
    IntegratedGradients,
    KernelShap,
    Lime,
    Occlusion,
    Saliency,
    SmoothGrad,
    Sobol,
    plot_attributions,
)
from interpreto.attributions import InferenceModes
from interpreto.attributions.metrics import Deletion, Insertion
from interpreto.commons import GranularityAggregationStrategy

# Instantiate the Occlusion explainer with the model and tokenizer
explainer = Occlusion(model, tokenizer)

# Compute the attributions on a given sentence
attributions = explainer(df.iloc[0]['text'])
```

Image

### Environment

* OS: 

Image

* Python version : Python 3.11.14
* Interpreto version : 0.4.19

## Expected behavior

The explanation

### PR DESCRIPTION
**Pull request by AntoninPoche**
This PR should be reviewed commit by commit.

### Initial modifications

This pull request aims at solving issue #137. To do so, when a model tokenizer has a `mask_token_id`, we use it as our `replace_token_id`, which prevents the dangerous `model.resize_token_embeddings()`.

Additionally, the tokenizer setting is abstracted so that both the `AttributionExplainer` and `InsertionDeletion` classes can use it.

**HOWEVER**, this has a significant consequence: the default replacement values change for all methods. It should not impact much, but we will not get the exact same attribution values. Therefore, the non-regression metric test failed. They were removed, but sanity checks were introduced.

Finally, I did not reproduce the bug in our test because I did not find any `hf-internal-testing/tiny-random-?` which would raise the error. As we do not want big models in our tests, I preferred not to put any. But we can decide otherwise.

### Some additional modification

- Build workflow fix [48717c9](https://github.com/FOR-sight-ai/interpreto/pull/138/commits/48717c9dfd50ad7bcf63bc36260d852dcda30ee9)
- Remove warning from `sentencepiece` incompatibility [7a1fb99](https://github.com/FOR-sight-ai/interpreto/pull/138/commits/7a1fb99cf1c7ff83d3bd932991af480e98f3a760)
- Add sanity checks. We modified the default attribution values; it was necessary. [4c2028a](https://github.com/FOR-sight-ai/interpreto/pull/138/commits/4c2028a970fb6135715f62976be6c02235b19fd9)
- Correct Sobol, it always had bad results, now we know why, it was bugged. This PR corrects it. [7fc90df](https://github.com/FOR-sight-ai/interpreto/pull/138/commits/7fc90dfef371ffbd8c587b2030d511258700a608)

**Note**: the sanity checks are marked as slow; thus, they are not run by default in the workflows.

## Related Issue

#137 showed that previously, the forced resizing caused the perturbation methods to break on `"camembert-base"`. It was due to the `model.resize_token_embeddings()`, which was mandatory.

## Type of Change

- 🔧 Bug fix (non-breaking change which fixes an issue)
- 🥂 Improvement (non-breaking change which improves an existing feature)
- 💥 Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: tests/model_wrapping/test_model_with_split_points.py [hunk 1]
```diff
@@ -669,6 +669,8 @@ def test_activations_and_gradients_on_models_long(model_name, sentences: list[st
 
 def evaluate_activations_and_gradients(model_name, sentences: list[str]):
     """Tests model with split points get activations and gradients with a large variety of models"""
+    if model_name == "hf-internal-testing/tiny-random-Gemma3ForCausalLM":
+        pytest.importorskip("sentencepiece")
```
Comment (fanny-jourdan) [inline tests/model_wrapping/test_model_with_split_points.py:673]: I dont understand this (i understand that skip the import but why?)

File: interpreto/attributions/base.py [hunk 2]
```diff
@@ -206,23 +234,7 @@ def __init__(
...
-
-        # resize model with new tokens
-        model.resize_token_embeddings(len(self.tokenizer))
-
-        replace_token_id = self.tokenizer.convert_tokens_to_ids(replace_token)
-        if isinstance(replace_token_id, list):
-            replace_token_id = replace_token_id[0]
-
-        return model, replace_token_id
+        return setup_tokenizer_for_perturbations(model, self.tokenizer)
```
Comment (fanny-jourdan) [inline interpreto/attributions/base.py:237]: Why not just write the code for the `setup_tokenizer_for_perturbations` function here instead of creating the function and calling it there?

File: tests/model_wrapping/test_model_with_split_points.py (hunk 1 above)
Comment (AntoninPoche) [reply tests/model_wrapping/test_model_with_split_points.py:673]: Actually, it is the other way around.

`"hf-internal-testing/tiny-random-Gemma3ForCausalLM"` requires `sentencepiece` to be installed; otherwise, it breaks. But because it is the only piece of code in the whole repo, I deemed unecessary to have `sentencpiece` as a main dependency.

So `pytest.importorskip` does not skip the import but the test if the import is not possible.

File: interpreto/attributions/base.py (hunk 2 above)
Comment (AntoninPoche) [reply interpreto/attributions/base.py:237]: It was the case before, but `InsertionDeletionBase` uses it too; thus, previously, the code was duplicated.

This code might cause a few bugs in the future, so it is better to have it in a single spot, so we do not forget to update the second version.

File: tests/model_wrapping/test_model_with_split_points.py (hunk 1 above)
Comment (fanny-jourdan) [reply tests/model_wrapping/test_model_with_split_points.py:673]: Oh, okay, thanks for clarifying!

File: tests/attributions/test_methods_sanity.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+if __name__ == "__main__":
+    model = AutoModelForSequenceClassification.from_pretrained(MODEL_ID).to(DEVICE)
+    tokenizer = AutoTokenizer.from_pretrained(MODEL_ID, use_fast=True)
+    model.eval()
+    test_classification_attribution_methods_detect_emotion_word((model, tokenizer, None), Sobol)
+    # tokenizer = LetterTokenizer()
+    # model = ShiftCopyForCausalLM(vocab_size=len(tokenizer)).to(DEVICE)
+    # model.eval()
+    # test_generation_attribution_methods_detect_copied_token((model, tokenizer), Sobol)
```
Comment (fanny-jourdan) [inline tests/attributions/test_methods_sanity.py:378]: You don't test generation finally ?

Comment (AntoninPoche) [reply tests/attributions/test_methods_sanity.py:378]: The lines in the `if __name__ == "__main__":` are not executed by `pytest`, so commenting these lines does not impact what is tested.

These lines are only executed if we directly run the script in which these lines are present. You could read the if statement `if __name__ == "__main__":` as "if the current file is the main executed file".

I use them often to debug, so I can specify which test I want to run and activate the debugger for it. This is why several lines are often commented out in this part.

**Review by fanny-jourdan — APPROVED** [review]
Thanks Antonin ! Everything is clear, good job ! 👍
